### PR TITLE
Set the default charset at startup

### PR DESCRIPTION
**Pull request by gregwebs**
### What problem does this PR solve?

This adds MySQL compatibility for setting a default charset (and collation).
To some extent it is just a work around for the lack of charset support that makes existing applications that set utf8mb4 fail. Here is a [bug report when someone tried to store data from Grafana](https://github.com/grafana/grafana/issues/12058).

I am a unsure if the current approach is good, so I haven't followed through with writing tests yet. I would be happy if the TiDB team took this over and used a better approach.

### What is changed and how it works?

I am now running tidb with the arguments:

    -character-set-server utf8mb4
    -collation-server utf8mb4_unicode_ci

My understaning is that this doesn't change the behavior of TiDB (it is
equivalent to the current configuration).
However, it allows TiDB to be compatible with standard MySQL setups.
Previously, I would see this failure message when using an application:

    unsupported modify column charset utf8 not match origin utf8mb4

Using the above command line flags, the application now works.
There is still more to do to have proper charset support,
but this is enough to fix my use case.

### Check List

Tests 

 - [ ] Unit test
 - [ ] Integration test
 - [X] Manual test

As per the above description, I manually tested that this fixed the issue with the application I am running (prisma).

Code changes

This change creates backwards compatible additions/enhancements.

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: sessionctx/variable/sysvar.go [hunk 1]
```diff
@@ -164,7 +164,7 @@ var defaultSysVars = []*SysVar{
 	{ScopeGlobal, "innodb_max_undo_log_size", ""},
 	{ScopeGlobal | ScopeSession, "range_alloc_block_size", "4096"},
 	{ScopeGlobal, ConnectTimeout, "10"},
-	{ScopeGlobal | ScopeSession, "collation_server", charset.CollationUTF8},
+	{ScopeGlobal | ScopeSession, "collation_server", mysql.DefaultCollationName},
```
Comment (lysu) [inline sessionctx/variable/sysvar.go]: I found `charset.CharsetUTF8, charset.CollationUTF8` was refered in another pkg, e.g. `expression/ builtin_*`, does them also need change?

**Comment by lysu** [comment]
@gregwebs the question that we want to solve is `could not modify column charset  from utf8 to utf8mb4`?

how abount only modify `modify column` ddl logic to open a door that can change utf8 to utf8mb4(because they are same in tidb and golang)?

this PR seems open a door to modify global charset configuration, user maybe use this to set `gbk, latin and other charset`, for current TiDB's situation support multiple charset is not easy(for utf8mb4, tidb just see it same as utf8)

maybe need PTAL @winkyao

**Comment by gregwebs** [comment]
I think making a change in `modify column` might be a hack and that the real way to fix this is to accept the charset setting somewhere where it is not being accepted on the initial schema creation (I am not sure where exactly it is not working right now: it may be the database, table, or column level). If you can outline how we would support setting the charset on schema creation we can look at making those changes.

Otherwise it is simple to validate that only utf8 and utf8mb4 are the allowed charsets for this setting: I can push a commit for that.

File: sessionctx/variable/sysvar.go (hunk 1 above)
Comment (gregwebs) [reply sessionctx/variable/sysvar.go]: I changed it in more places now :)

File: expression/builtin_math_test.go [hunk 2]
```diff
@@ -557,7 +557,7 @@ func (s *testEvaluatorSuite) TestConv(c *C) {
 		tp := f.GetType()
 		c.Assert(tp.Tp, Equals, mysql.TypeVarString)
 		c.Assert(tp.Charset, Equals, charset.CharsetUTF8)
-		c.Assert(tp.Collate, Equals, charset.CharsetUTF8)
+		c.Assert(tp.Collate, Equals, charset.CollationUTF8)
```
Comment (gregwebs) [inline expression/builtin_math_test.go]: This is a bugfix from this PR, although I am not sure what part of this changset fixed it.

**Comment by lysu** [comment]
@gregwebs if we start tidb-server with

```
-character-set-server utf8mb4
-collation-server utf8mb4_unicode_ci
```

can we change `utf8mb4` back to `utf8`?

**Comment by coocood** [comment]
@gregwebs 
I think the change is too much relative to the problem it solves.
Support modify column charset from utf8 to utf8mb4 is much simpler.

**Comment by gregwebs** [comment]
@lysu no, you cannot change it back. That would be an additional feature to make the global variables work, but hopefully an easy change we can add on. Maybe that is needed to make this PR look right?

Right now this changeset is focused on being good enough to solve the most common real-world usage where utf8mb4 is used for everything. We could also solve this by just switching our default charset from utf8 -> utf8mb4, is there a reason why we default to utf8?

@coocood can you outline the changes to support modify column? But after that it is quite possible for there to be confusion and bugs because the real problem is that the column was supposed to be created with utf8mb4, not utf8. The modify issue is just an artifact of that. But it is useful to properly support modify, and I can test to see if that solves my problem for now.

I agree this PR touches a lot for solving just my issue. Maybe you are thinking of a more elegant way of doing it? However, I think this change is necessary at some point to have proper MySQL compatibility, particularly if we add some of the related global variable setting.

**Comment by morgo** [comment]
This will also become an issue for MySQL 8.0 compatibility, where utf8mb4 (not utf8) is the default character set.  Some applications expect/depend on utf8mb4.

**Comment by gregwebs** [comment]
Maybe instead of flags for setting charset, we should have `--compatibility 8`.

**Comment by morgo** [comment]
Where possible, I think it's better to target a new GA as 8.0 compatible.  Having modes can hurt usability, and users rarely change from defaults.

**Comment by gregwebs** [comment]
I think we should separate UX from internal implementation. Internally, I doubt think TiDB can afford to have to backport fixes to an old 5.7 only release and we will want one system that supports both modes, particularly as we start to add 8.0 compatibility but have not yet completed it.

This could be exposed to the user via a single config option (compatibility mode), a collection of config options (config files), or two different binary builds with different default settings built-in. With any of these options, at some point we can switch to 8.0 as default (in the binary case that is more about naming the binaries).

File: executor/ddl.go [hunk 3]
```diff
@@ -174,11 +174,11 @@ func (e *DDLExec) executeDropDatabase(s *ast.DropDatabaseStmt) error {
 	sessionVars := e.ctx.GetSessionVars()
 	if err == nil && strings.ToLower(sessionVars.CurrentDB) == dbName.L {
 		sessionVars.CurrentDB = ""
-		err = variable.SetSessionSystemVar(sessionVars, variable.CharsetDatabase, types.NewStringDatum("utf8"))
+		err = variable.SetSessionSystemVar(sessionVars, variable.CharsetDatabase, types.NewStringDatum(mysql.DefaultCharset))
 		if err != nil {
 			return errors.Trace(err)
 		}
-		err = variable.SetSessionSystemVar(sessionVars, variable.CollationDatabase, types.NewStringDatum("utf8_unicode_ci"))
```
[the remark this answers is not in the document]
Comment (jackysp) [reply executor/ddl.go]: I don't think so.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -122,7 +122,7 @@ func getDefaultCharsetAndCollate() (string, string) {
 	// TODO: TableDefaultCharset-->DatabaseDefaultCharset-->SystemDefaultCharset.
```
Comment (jackysp) [inline ddl/ddl_api.go]: I think we need to finish this TODO to solve the issue in the pr description.
We could return the values of "character_set_server", or "character_set_database" here.
Then we could run the SQL `set global/session character_set_server = 'utf8mb4'` to change the default character set.

File: executor/ddl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jackysp) [reply executor/ddl.go]: Yes, I think it's a good change.

File: ddl/ddl_api.go (hunk 4 above)
Comment (gregwebs) [reply ddl/ddl_api.go]: Right, this TODO issue you bring up could have provided me with alternative ways to try to fix this issue. Alter the database default charset probably would have worked, but would require altering my workflow.

I just reviewed the variable interaction of this PR more and pushed a commit to revert the change to `character_set_system`.

I agree in general with the philosophy that configuration should be update-able at runtime. However, in this case
1) there is no variable matching the default charset in MySQL, it is just a [startup or compile-time configuration](https://dev.mysql.com/doc/refman/8.0/en/charset-server.html)
2) it creates more code to maintain (with potential for race conditions)
3) I don't think there is a valid use case for changing the default charset at runtime. The only use case would seem to be recognizing a mis-configuration issue that you want to correct by changing once, which can be done via restart. For performance tuning variables, for example, you don't know what the correct configuration is and you have to tinker.

Let me know if that makes sense.

File: tidb-server/main.go [hunk 5]
```diff
@@ -337,7 +352,7 @@ func overrideConfig() {
 	}
 }
 
-func validateConfig() {
+func validateConfig() error {
```
Comment (jackysp) [inline tidb-server/main.go]: I think It's better to print the error log then exit instead of return an error.

File: ddl/ddl_api.go (hunk 4 above)
Comment (jackysp) [reply ddl/ddl_api.go]: OK.

File: tidb-server/main.go [hunk 6]
```diff
@@ -371,9 +386,11 @@ func validateConfig() {
 		log.Errorf("lower-case-table-names should be 0 or 1 or 2.")
 		os.Exit(-1)
 	}
+
+	return mysql.ValidateDefaultCharsetCollation(cfg.CharacterSetServer, cfg.CollationServer)
 }
 
-func setGlobalVars() {
+func setGlobalVars() error {
```
Comment (jackysp) [inline tidb-server/main.go]: We can also avoid this return value.

File: tidb-server/main.go (hunk 5 above)
Comment (gregwebs) [reply tidb-server/main.go]: Its really problematic to have exits anywhere other then the main function. The linters warn against this. In this case it is a private function in main.go, but still any function which exits is not testable.

I agree that being in an intermediate state where some return error and some still exit is not great, but I don't want to start cleaning up other code in this PR. Although it is still possible I might be forced to when I start writing test cases.

Comment (jackysp) [reply tidb-server/main.go]: OK, I see. But it still looks so strange. We need to clean up the other code anyway, but may not in this PR. It's better to add a TODO here.

Comment (gregwebs) [reply tidb-server/main.go]: okay

**Comment by gregwebs** [comment]
I am still hoping for agreement that this PR can be merged once the quality is good enough. Once agreed, I can write tests (write down any specific test cases you have in mind here).

**Comment by jackysp** [comment]
It seems there is no need to modify the ddl package, so I remove the label `component/DDL-need-LGT3`.

**Comment by gregwebs** [comment]
I will remove the command line option to change the default charset. We need to just switch the default to utf8mb4, which is the correct MySQL terminology.

File: expression/builtin_control.go [hunk 7]
```diff
@@ -81,19 +80,19 @@ func inferType4ControlFuncs(lhs, rhs *types.FieldType) *types.FieldType {
 			}
 		}
 		if types.IsNonBinaryStr(lhs) && !types.IsBinaryStr(rhs) {
-			resultFieldType.Charset, resultFieldType.Collate, resultFieldType.Flag = charset.CharsetUTF8, charset.CollationUTF8, 0
+			resultFieldType.Charset, resultFieldType.Collate, resultFieldType.Flag = mysql.DefaultCharset, mysql.DefaultCollationName, 0
```
Comment (winkyao) [inline expression/builtin_control.go]: No need to change to mysql.DefaultCharset, mysql.DefaultCollationName. This type is string, so the charset is definite utf8mb4(after my pr merged.)

Comment (winkyao) [inline expression/builtin_control.go]: No need to change to mysql.DefaultCharset, mysql.DefaultCollationName. This type is string, so the charset is definite utf8mb4(after my pr merged.) and my PR will change here to `charset.CharsetUTF8MB4, charset.CollationUTF8MB4`

File: expression/builtin_math_test.go (hunk 2 above)
Comment (winkyao) [reply expression/builtin_math_test.go]: My pr also fixed it... :)

**Comment by winkyao** [comment]
@gregwebs I will support change any charset to utf8 or utf8mb4 later, I think it will solve your problem.

**Comment by winkyao** [comment]
@gregwebs I will support change any charset to utf8 or utf8mb4 later, I think it will solve your problem.

**Comment by winkyao** [comment]
@gregwebs I will support change any charset to utf8 or utf8mb4 later, I think it will solve your problem.

**Comment by gregwebs** [comment]
I am just going to close this for now. We can revisit this if we have a concrete use case for collation support, etc.

**Pull request closed without merging**